File: scripts/make-manifest.php
<?php

declare(strict_types=1);

$dist = $argv[1] ?? __DIR__.'/../dist';
$dist = realpath($dist);

if ($dist === false) {
    fwrite(STDERR, "Distribution directory does not exist.\n");
    exit(1);
}

$assets = [];
$checksums = [];

foreach (glob($dist.'/privacy-filter-*.{tar.gz,zip}', GLOB_BRACE) ?: [] as $path) {
    $filename = basename($path);

    if (! preg_match('/^privacy-filter-(?<os>linux|darwin|windows)-(?<arch>x64|arm64)\.(?<extension>tar\.gz|zip)$/', $filename, $matches)) {
        continue;
    }

    $sha256 = hash_file('sha256', $path);

    $assets[] = [
        'os' => $matches['os'],
        'arch' => $matches['arch'],
        'cpuOnly' => true,
        'filename' => $filename,
        'sha256' => $sha256,
    ];

    $checksums[] = "{$sha256}  {$filename}";
}

usort($assets, fn (array $a, array $b): int => [$a['os'], $a['arch']] <=> [$b['os'], $b['arch']]);
sort($checksums);

$manifest = [
    'schemaVersion' => 1,
    'assets' => $assets,
];

file_put_contents($dist.'/manifest.json', json_encode($manifest, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES).PHP_EOL);
file_put_contents($dist.'/checksums.txt', implode(PHP_EOL, $checksums).PHP_EOL);

File: scripts/make-manifest.py
#!/usr/bin/env python3
import hashlib
import json
import re
import sys
from pathlib import Path


def sha256(path):
    digest = hashlib.sha256()

    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)

    return digest.hexdigest()


def main():
    dist = Path(sys.argv[1] if len(sys.argv) > 1 else Path(__file__).resolve().parent.parent / "dist")

    if not dist.is_dir():
        raise SystemExit("Distribution directory does not exist.")

    pattern = re.compile(r"^privacy-filter-(?P<os>linux|darwin|windows)-(?P<arch>x64|arm64)\.(?P<extension>tar\.gz|zip)$")
    archive_paths = sorted([*dist.glob("privacy-filter-*.tar.gz"), *dist.glob("privacy-filter-*.zip")])
    assets = []
    checksums = []

    for path in archive_paths:
        match = pattern.match(path.name)

        if not match:
            continue

        file_hash = sha256(path)

        assets.append({
            "os": match.group("os"),
            "arch": match.group("arch"),
            "cpuOnly": True,
            "filename": path.name,
            "sha256": file_hash,
        })

        checksums.append(f"{file_hash}  {path.name}")

    assets.sort(key=lambda asset: (asset["os"], asset["arch"]))
    checksums.sort()

    manifest = {
        "schemaVersion": 1,
        "assets": assets,
    }

    (dist / "manifest.json").write_text(json.dumps(manifest, indent=4) + "\n", encoding="utf-8")
    (dist / "checksums.txt").write_text("\n".join(checksums) + "\n", encoding="utf-8")


if __name__ == "__main__":
    main()
